### DISCUSSION TITLE
高级搜索的默认条件可以允许自定义就好了

### DISCUSSION THREAD

**Discussion by WeiJunFenYou** [dotnetcore/BootstrapBlazor#4170] [Ideas]
在Table组件中，其自动生成的搜索，QueryPageOptions的AdvanceSearches属性，表达式的FilterAction被硬编码为
`FilterAction.Equal`，可以给TableColumn新增一个属性“FilterAction”，在Filterable为true时，读取FilterAction的值，默认为FilterAction.Equal。这样用户在构建搜索框时，需要用到“包含”、“大于等于”等等关系符时，就不需要兴师动众自己再定义一个模板了。

[代码链接](https://github.com/dotnetcore/BootstrapBlazor/blob/20f7bc7e7ed89b0d7bd022b5602126aa31707148/src/BootstrapBlazor/Components/Table/Table.razor.Search.cs#L257)

**Comment by ArgoZhang** [comment]
@WeiJunFenYou 感谢提交讨论，高级搜索是一个表单，里面有多个条件，如果这里不是等于的条件，我个人觉得应该是有提示让使用者知道是大于或者小于或者其他条件，不然这让用户非常困扰。所以内置的就是等于这条件，符合常识。如果有更好的解决方案我们非常乐意接受。请提出宝贵建议与解决方案，我们可以一起讨论

**Reply by WeiJunFenYou** [reply]
未必如此吧？搜索**姓名、住址、备注**等等字符串文本信息，一般都是“包含”关系。极少有人会用“等于”关系，如此会极不实用，搜索文本但凡差错一字，都搜不到想要的信息。搜索文本时，“包含”关系才是符合直觉的。但在搜索手机号时，如果手机号是以字符串形式储存的，就又需要“等于”关系。

而搜索数值类型的信息，例如枚举、年龄，才会默认使用“等于”。

鄙人正在使用本库构建一个应用，恰好用了高级搜索功能。搜索文本的条件正是“包含”关系，正是发现AdvanceSearches属性里的默认关系是`FilterAction.Equal`，且无法改变，才会有此疑问。为了这么一个小功能就自定义一个搜索模板实在大炮打蚊子。但眼下没有更好的办法，也只能这么做了。

**Reply by ArgoZhang** [reply]
@WeiJunFenYou 你的想法我也都理解。所以我前面提到了。如果你是模糊查询，我还是建议有一个提示，这个条件是模糊查询，不能说所有查询条件用模糊的就合理，也不能说所有的查询条件用精确匹配就合理。这完全是业务问题，组件库只是帮你做了 UI 而已，如果不用我们的组件库，这个 UI 是需要你自己做的，仅此而已。你有什么好的想法，好的解决方案可以提，如果合理、通用，你提交代码也可以，给我们思路，我们做也可以。

**Comment by WeiJunFenYou** [comment]
内置的搜索条件是“等于”没有太大问题，但还是建议本库给用户选择的空间，可以默认“等于”，但允许用户可以自己配置成“包含、大于、小于”的条件。至于提示信息，一段页面上的文字注释即可。

**Reply by ArgoZhang** [reply]
所以怎么给？怎么让用户去设置？能否提供切实可行的解决方案？

**Comment by WeiJunFenYou** [comment]
`<Table TItem="Dto"
       IsPagination="true" PageItemsSource="@PageItemsSource" SearchDialogShowMaximizeButton="true"
       IsStriped="true" IsBordered="true" SearchModel="@SearchModel"
       ShowToolbar="true" ShowDefaultButtons="false" IsMultipleSelect="true" ShowExtendButtons="true"
       ShowSearch="true" ShowAdvancedSearch="false" OnQueryAsync="@OnQueryAsync">
    <TableColumns>
        <TableColumn @bind-Field="@context.Type1" Text="类别1"  />
        <TableColumn @bind-Field="@context.Type2" Text="类别2" Searchable="true" />
        <TableColumn @bind-Field="@context.Type3" Text="类别3" Searchable="true" />
        <TableColumn @bind-Field="@context.Type4" Text="类别4" />
        <TableColumn @bind-Field="@context.Type5" Text="类别5" />
    </TableColumns>
</Table>`

例如又这么一段代码，“类别2”和“类别3”使用了`Searchable="true"`，在高级搜索框中会显示出来。可否增加一个属性，就叫“FilterAction”，让用户自己指定。
        `<TableColumn @bind-Field="@context.Type3" Text="类别3" Searchable="true"  FilterAction="FilterAction.Contains" />`

**Reply by WeiJunFenYou** [reply]
@ArgoZhang
